### PR TITLE
Add metric producers to meter_provider configuration

### PR DESCRIPTION
**Pull request by dashpole**
Part of https://github.com/open-telemetry/opentelemetry-specification/issues/4015

Metric producers are registered to metric readers, so i've nested it under the periodic and manual readers.

@gouthamve @jack-berg

### CODE REVIEW

File: schema/meter_provider.json [hunk 1]
```diff
@@ -75,6 +87,27 @@
                 }
             }
         },
+        "MetricProducer": {
+            "type": "object",
+            "additionalProperties": true,
+            "minProperties": 1,
+            "maxProperties": 1,
+            "properties": {
+                "opencensus": {
```
Comment (dashpole) [inline schema/meter_provider.json:96]: This was the only part I wasn't entirely sure about.  Since this is a pluggable interface, i've made them objects, with additional properties allowed.  I think that would allow adding configuration to these readers in the future.

File: schema/meter_provider.json [hunk 2]
```diff
@@ -75,6 +87,27 @@
...
+                    "type": "object",
+                    "additionalProperties": false
+                },
+                "prometheus": {
```
Comment (dashpole) [inline schema/meter_provider.json:100]: I wasn't sure whether to include the prometheus bridge here.  AFAIK this only exists in Go (maybe java too?).  Are properties added here required for implementations?  Or are they optional, and just a way to standardize the keys?

File: schema/meter_provider.json [hunk 3]
```diff
@@ -33,6 +33,12 @@
                 },
                 "exporter": {
                     "$ref": "#/$defs/MetricExporter"
+                },
+                "producers": {
```
Comment (jack-berg) [inline schema/meter_provider.json:37]: Let's add this as a property on `MetricReader` instead of PeriodicMetricReader and PullMetricReader.

File: schema/meter_provider.json (hunk 2 above)
Comment (jack-berg) [reply schema/meter_provider.json:100]: We only add schema surface area for components that are explicitly defined in the spec. The go implementation of this producer should be registered as a [ComponentProvider](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/configuration/file-configuration.md#component-provider).

File: schema/meter_provider.json (hunk 1 above)
Comment (jack-berg) [reply schema/meter_provider.json:96]: As mentioned in another comment, we only add schema surface area for components which are explicitly definitely in the spec. Opencensus meets this critiera: https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/compatibility/opencensus.md#metrics--stats

Comment (dashpole) [reply schema/meter_provider.json:90]: OpenCensus is now the only one in this list.

File: schema/meter_provider.json (hunk 2 above)
Comment (dashpole) [reply schema/meter_provider.json]: I removed this component

File: schema/meter_provider.json (hunk 3 above)
Comment (dashpole) [reply schema/meter_provider.json]: Moved

File: schema/meter_provider.json [hunk 4]
```diff
@@ -110,6 +133,12 @@
                 },
                 "pull": {
                     "$ref": "#/$defs/PullMetricReader"
+                },
+                "producers": {
```
Comment (jack-berg) [inline schema/meter_provider.json:137]: Looks like we need to remove the `maxProperties` field. The goal for this was to constrain you to either specify pull or periodic, but not both. But with another property at the same level, we need to drop this constraint (or find a new way to enforce it).

File: examples/kitchen-sink.yaml [hunk 5]
```diff
@@ -142,6 +142,10 @@ meter_provider:
               # Configure resource attributes to be excluded, in this example attribute service.attr1.
               excluded:
                 - "service.attr1"
+        # Configure metric producers.
+        producers:
```
Comment (jack-berg) [inline examples/kitchen-sink.yaml:146]: I believe this is at the wrong level. (Applies to the other usage as well)
```suggestion
      producers:
```

File: schema/meter_provider.json (hunk 4 above)
Comment (dashpole) [reply schema/meter_provider.json:136]: removed

File: examples/kitchen-sink.yaml (hunk 5 above)
Comment (dashpole) [reply examples/kitchen-sink.yaml]: fixed here and below

File: examples/kitchen-sink.yaml [hunk 6]
```diff
@@ -142,6 +142,10 @@ meter_provider:
               # Configure resource attributes to be excluded, in this example attribute service.attr1.
               excluded:
                 - "service.attr1"
+        # Configure metric producers.
```
Comment (codeboten) [inline examples/kitchen-sink.yaml:145]:
```suggestion
      # Configure metric producers.
```

File: examples/kitchen-sink.yaml [hunk 7]
```diff
@@ -203,6 +207,10 @@ meter_provider:
             #
             # Environment variable: OTEL_EXPORTER_OTLP_METRICS_DEFAULT_HISTOGRAM_AGGREGATION
             default_histogram_aggregation: base2_exponential_bucket_histogram
+        # Configure metric producers.
```
Comment (codeboten) [inline examples/kitchen-sink.yaml:210]:
```suggestion
      # Configure metric producers.
```

File: schema/meter_provider.json [hunk 8]
```diff
@@ -103,13 +126,18 @@
             "type": "object",
             "additionalProperties": false,
             "minProperties": 1,
-            "maxProperties": 1,
```
Comment (dashpole) [inline schema/meter_provider.json:106]: Should I keep this, and just make it `"maxProperties": 2`

Comment (jack-berg) [reply schema/meter_provider.json:106]: That seems fine for now.

**Review by jack-berg — APPROVED** [review]

File: schema/meter_provider.json (hunk 8 above)
Comment (dashpole) [reply schema/meter_provider.json:106]: As in: It is fine to change it to `"maxProperties": 2`, or the current state is fine?

Comment (jack-berg) [reply schema/meter_provider.json:106]: Adding `maxProperties: 2` seems fine even if it doesn't perfectly provide the constraint we want.

Comment (dashpole) [reply schema/meter_provider.json:106]: added

**Review by codeboten — APPROVED** [review]

File: schema/meter_provider.json [hunk 9]
```diff
@@ -46,6 +46,12 @@
             "properties": {
                 "exporter": {
                     "$ref": "#/$defs/MetricExporter"
+                },
+                "producers": {
+                    "type": "array",
+                    "items": {
+                        "$ref": "#/$defs/MetricProducer"
+                    }
                 }
```
Comment (bogdandrutu) [inline schema/meter_provider.json:55]: Why is this needed here if already in the top level MetricReader? If it is required for pull (maybe to control at a granular level) why not for periodic as well?

Comment (dashpole) [reply schema/meter_provider.json:55]: https://github.com/open-telemetry/opentelemetry-configuration/pull/93

Comment (jack-berg) [reply schema/meter_provider.json:55]: Thanks for calling this out @bogdandrutu!
